### PR TITLE
Optimize chunk extraction

### LINKED ISSUE #275: carve_chunk_to_file use sendfile instead of a read/wrtite loop
read write loop is slow and resource intensive, we could use os.sendfile

low-prio

### PR DESCRIPTION
**Pull request by e3krisztian**
fixes #275

### CODE REVIEW

File: tests/test_extractor.py [hunk 1]
```diff
@@ -20,21 +20,29 @@ def test_no_chunks(self, tmp_path: Path):
 
     def test_one_chunk(self, tmp_path: Path):
         content = b"test file"
-        test_file = io.BytesIO(content)
+        infile = tmp_path / "infile"
```
Comment (kissgyorgy) [inline tests/test_extractor.py:23]: When you change implementation, but the feature is the same, tests should not be changed at all, which would show that the features we put there in the first place still work.

File: tests/test_extractor.py [hunk 2]
```diff
@@ -20,21 +20,29 @@ def test_no_chunks(self, tmp_path: Path):
...
+
         written_path = tmp_path / "0-9.unknown"
-        assert list(tmp_path.iterdir()) == [written_path]
+        assert set(tmp_path.iterdir()) == {infile, written_path}
         assert written_path.read_bytes() == content
 
     def test_multiple_chunks(self, tmp_path: Path):
         content = b"test file"
-        test_file = io.BytesIO(content)
+        infile = tmp_path / "infile"
```
Comment (kissgyorgy) [inline tests/test_extractor.py:36]: Same as above, tests should not be touched, when the feature doesn't change.

**Review by kissgyorgy — CHANGES_REQUESTED** [review]
The commit message doesn't say anything about WHY the interface has been broken. Please follow our developer guide for commit messages!

The reason using `io.BufferedIOBase` everywhere is not by accident. We wanted to make the codebase very flexible, because the files could come from anywhere. For example, it would be possible to download a file with a HTTP request and pass the response to the `carve_chunk_to_file` function, not having to save the file to disk. OR what's even more important don't have to create actual files for tests and use `io.BytesIO` instead, to make the test suite faster (at least the unit tests).

I don't see the *intent* of breaking this (the commit message doesn't reflect this decision and the type hints has not been changed), I think you just broke it by accident. I think we should keep this flexibility, so `carve_chunk_to_file` should handle `io.BufferedIOBase` still, not just ordinary file types (`typing.BinaryIO`).  
We can break this interface, but it has to be discussed and the reason has to be documented if we chose to do so.

**Comment by vlaci** [comment]
I just wanted to chime in saying that I would be impressed if yara could match on e.g a BytesIO object so that I wouldn't be overly optimistic about having these possibilities at all.

File: unblob/extractor.py [hunk 3]
```diff
@@ -31,8 +30,7 @@ def make_extract_dir(root: Path, path: Path, extract_root: Path) -> Path:
 def carve_chunk_to_file(carve_path: Path, file: io.BufferedIOBase, chunk: Chunk):
     """Extract valid chunk to a file, which we then pass to another tool to extract it."""
     with carve_path.open("wb") as f:
-        for data in iterate_file(file, chunk.start_offset, chunk.size):
-            f.write(data)
+        os.sendfile(f.fileno(), file.fileno(), chunk.start_offset, chunk.size)
```
[the remark this answers is not in the document]
Comment (vlaci) [reply unblob/extractor.py:33]: That's odd,
```
>>> import io
>>> io.BufferedIOBase
<class 'io.BufferedIOBase'>
```

The method will raise an `io.UnsupportedOperation` exception when there is no underlying file object.

[the remark this answers is not in the document]
Comment (e3krisztian) [reply unblob/extractor.py:33]: Yes, indeed this was the reason, the interface was changed to support only objects with `.fileno()`.

The proper (both optimization and keep working without `fileno`) implementation would check if there is a `fileno` and if not, go with the original implementation.

If we want to support `BytesIO`, I do not think this optimization is worth it: we will have a more complex code and marginal (if any) improvement in some cases.

File: tests/test_extractor.py (hunk 1 above)
Comment (e3krisztian) [reply tests/test_extractor.py:23]: agreed.

**Comment by e3krisztian** [comment]
Discussed in person, this optimization is not worth the changes.

**Pull request closed without merging**